src/main/scala/com/fulcrumgenomics/bam/FindTechnicalReads.scala

```diff
@@ -0,0 +1,113 @@
+import com.fulcrumgenomics.util.{IlluminaAdapters, Io, ProgressLogger}
+import dagr.commons.util.LazyLogging
+import dagr.sopt._
+import htsjdk.samtools.util.SequenceUtil
+import htsjdk.samtools.{SAMFileWriterFactory, SamReaderFactory}
+
+import scala.annotation.tailrec
+import scala.collection.JavaConversions.asScalaIterator
+
+@clp(group = ClpGroups.SamOrBam, description =
+  """
+     |Find reads that are from technical or synthetics sequences in a BAM file. Takes in
+     |a BAM file, extracts the read pairs and fragment reads that are unmapped, and tests
+     |them to see if they are likely generated from a technical sequence (e.g. adapter
+     |dimer).
+     |
+     |The identification of reads is done by testing the first N bases (controlled by the
+     |match-length parameter) of each read against all sub-sequences of length N from the
+     |technical sequences.  Sub-sequences are generated from both the sequences and the
+     |reverse complement of the sequences, ignoring any sub-sequences that include 'N's.
+     |
+     |The output BAM file will contain all reads that matched to a sub-sequence of the
+     |technical sequences and, if the read is paired, the read's mate pair.
+     |
+     |The default set of sequences include a range of different Illumina adapter sequences
+     |with the sample index/barcode region masked to Ns.
+  """
+  )
+class FindTechnicalReads
+( @arg(flag="i", doc="Input SAM or BAM file") val input: PathToBam,
+  @arg(flag="o", doc="Output SAM or BAM file") val output: PathToBam,
+  @arg(flag="m", doc="The number of bases at the start of the read to match against.") val matchLength: Int = 15,
+  @arg(flag="e", doc="The maximum number of errors in the matched region.") val maxErrors: Int = 1,
+  @arg(flag="s", doc="The set of technical sequences to look for.") val sequences: Seq[String] = IlluminaAdapters.all.flatMap(_.both)
+) extends FgBioTool with LazyLogging {
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+
+  override def execute(): Unit = {
+    val in       = SamReaderFactory.make().open(input.toFile)
+    val iter     = if (in.hasIndex) in.queryUnmapped().buffered
+                   else in.iterator().filter(r => r.getReadUnmappedFlag && (!r.getReadPairedFlag || r.getMateUnmappedFlag)).buffered
+    val out      = new SAMFileWriterFactory().makeWriter(in.getFileHeader, true, output.toFile, null)
+    val matcher  = new Matcher(sequences=sequences, matchLength=matchLength, maxErrors=maxErrors)
+    val progress = new ProgressLogger(logger, unit=250000)
+    var n: Long  = 0
+
+    while (iter.hasNext) {
+      val r1 = iter.next()
+      val r2 = if (iter.hasNext && iter.head.getReadName == r1.getReadName) Some(iter.next()) else None
+      val recs = Seq(Some(r1), r2).flatten
+      if (recs.exists(r => matcher.matches(r.getReadBases))) {
+        recs.foreach(out.addAlignment)
+        n += recs.size
+      }
+
+      recs.foreach(progress.record)
```

Comment:
does progress have a `counts` method that returns the # of records it has encountered?  If so, no need for `n`.